### PR TITLE
Fix a storage leak in parachains db

### PR DESCRIPTION
**Pull request by eskimor**
This PR makes sure old votes are pruned from the database in dispute-coordinator.

Should be tested on Versi and with paritydb.

- [x] Tested on Rococo - disk usage seems to [go down](https://grafana.parity-mgmt.parity.io/d/919b92a8e8041bd567af9edab12c840c/kubernetes-persistent-volumes?orgId=1&var-datasource=prometheus.parity-testnet&var-cluster=&var-namespace=rococo&var-volume=chain-data-rococo-validator-a-node-0&from=now-3h&to=now).
- [x] Tested on paritydb: Nodes are crashing, but things are working after a db reset:
```
 WARN tokio-runtime-worker sc_service::builder: The NetworkStart returned as part of `build_network` has been silently dropped    
Error: 
   0: InvalidConfiguration("Column config mismatch for column 4. Expected \"preimage: false, uniform: false, refc: false, compression: 0, ordered: true\", got \"preimage: false, uniform: false, refc: false, compression: 0, ordered: false\"")
```

Operators will need to delete the parachains paritydb.

- [ ] burnin on Kusama and check for saved storage space.

### CODE REVIEW

File: node/core/dispute-coordinator/src/db/v1.rs [hunk 1]
```diff
@@ -258,20 +323,116 @@ pub(crate) fn note_current_session(
...
+/// Until what session votes have been cleaned up already.
+///
+/// That is the db has already been purged of votes for sessions older than the returned
+/// `SessionIndex`.
+fn load_cleaned_votes_watermark(
+	db: &dyn Database,
+	config: &ColumnConfiguration,
+) -> FatalResult<Option<SessionIndex>> {
+	load_decode(db, config.col_data, CLEANED_VOTES_WATERMARK_KEY)
+		.map_err(|e| FatalError::DbReadFailed(e))
```
Comment (ordian) [inline node/core/dispute-coordinator/src/db/v1.rs:335]: why introduce a separate error type? other methods use
`SubsystemError::with_origin("dispute-coordinator", e)`

we already have `SubsystemError::IoError` and `Error::Io`, no reason to introduce yet another copy of that error

Comment (eskimor) [reply node/core/dispute-coordinator/src/db/v1.rs:335]: I needed it compatible with the `FatalResult` used in `write`. The errors here are a mixture of legacy and "new" error types, this should get cleaned up at some point indeed.

File: node/core/dispute-coordinator/src/db/v1.rs [hunk 2]
```diff
@@ -79,12 +99,50 @@ impl Backend for DbBackend {
 		for op in ops {
 			match op {
 				BackendWriteOp::WriteEarliestSession(session) => {
+					// Cleanup old votes in db:
+					let watermark =
```
Comment (sandreim) [inline node/core/dispute-coordinator/src/db/v1.rs:103]: Nit: I would move this into a separate fn like `periodic_vote_cleanup` to make the code more readable.

File: node/core/dispute-coordinator/src/db/v1.rs [hunk 3]
```diff
@@ -31,12 +31,30 @@ use crate::{
...
+///
+/// 300 is with session duration of 1h and 30 parachains around <3_000_000 key purges in the worst
+/// case. Which is already quite a lot, at the same time we have around 21_000 sessions on
+/// Kusama. This means at 300 purged sessions per session, cleaning everything up will take
+/// around 3 days. Depending on how severe disk usage becomes, we might want to bump the batch
+/// size, at the cost of risking issues at session boundaries (performance).
+#[cfg(test)]
+const MAX_CLEAN_BATCH_SIZE: u32 = 10;
+#[cfg(not(test))]
+const MAX_CLEAN_BATCH_SIZE: u32 = 300;
```
Comment (sandreim) [inline node/core/dispute-coordinator/src/db/v1.rs:57]: What is the impact of doing this many items on each earliest session update? (for nodes who have a lot of dangling storage items to clean)

**Review by sandreim — APPROVED** [review]
LGTM

File: node/core/dispute-coordinator/src/db/v1.rs (hunk 2 above)
Comment (eskimor) [reply node/core/dispute-coordinator/src/db/v1.rs:103]: I was on the edge ;-) Will do!

File: node/core/dispute-coordinator/src/db/v1.rs (hunk 3 above)
Comment (eskimor) [reply node/core/dispute-coordinator/src/db/v1.rs:57]: burnin on Kusama will tell.

File: node/core/dispute-coordinator/src/db/v1.rs [hunk 4]
```diff
@@ -31,12 +31,30 @@ use crate::{
...
+/// Until what session have votes been cleaned up already?
+const CLEANED_VOTES_WATERMARK_KEY: &[u8; 23] = b"cleaned-votes-watermark";
+
+/// Restrict number of cleanup operations.
+///
+/// On the first run we are starting at session 0 going up all the way to the current session -
+/// this should not be done at once, but rather in smaller batches so nodes won't get stalled by
+/// this.
+///
+/// 300 is with session duration of 1h and 30 parachains around <3_000_000 key purges in the worst
```
Comment (ordian) [inline node/core/dispute-coordinator/src/db/v1.rs:49]:
```suggestion
/// 300 is with session duration of 1 hour and 30 parachains around <3_000_000 key purges in the worst
```

File: node/core/dispute-coordinator/src/db/v1.rs [hunk 5]
```diff
@@ -47,6 +65,44 @@ impl DbBackend {
 	pub fn new(db: Arc<dyn Database>, config: ColumnConfiguration) -> Self {
 		Self { inner: db, config }
 	}
+
+	/// Cleanup old votes.
+	///
+	/// Should be called whenever a new ealiest session gets written.
```
Comment (ordian) [inline node/core/dispute-coordinator/src/db/v1.rs:71]:
```suggestion
	/// Should be called whenever a new earliest session gets written.
```

**Review by ordian — APPROVED** [review]
How did burn-in on Kusama go? What are the expected savings?

File: node/core/dispute-coordinator/src/db/v1.rs (hunk 5 above)
Comment (eskimor) [reply node/core/dispute-coordinator/src/db/v1.rs:71]: Hmm, I thought I fixed that already. Anyway, thanks!

**Comment by eskimor** [comment]
burnin on Kusama was not done yet. On Rococo I don't have definite data, but it looks like around 10%. I will add a metric how long cleanup rounds take and will then burnin on Kusama.

**Comment by Generic-Chain** [comment]
this has not been resolved, since the release of v0.9.25 (and after a full cleanup of parachains/db folder) it has gradually increased with .sst files left behind since Jul 6 every day. At the moment on a validator that is active it has gotten to 376 .sst files total and 23GB in size of the parachains/db folder.

# du -hs .local/share/polkadot/chains/ksmcc3/db/full/parachains/db/
24G     .local/share/polkadot/chains/ksmcc3/db/full/parachains/db/
# ls -ltr .local/share/polkadot/chains/ksmcc3/db/full/parachains/db/*.sst | wc -l
376

**Comment by eskimor** [comment]
Thanks @Generic-Chain ! Indeed there exists another leak, @vstakhov already identified that one as well and I believe there exists a fix already too. @vstakhov I was not in the loop recently, what is the status on this?

**Comment by vstakhov** [comment]
It is not very likely that the leak I have fixed could lead to 23Gb of storage waste. One thing we could try is to examine the current database to figure out what's happening.

**Comment by Generic-Chain** [comment]
I'm just reporting what I am seeing, I'm not familiar with what exactly was fixed and if there's some other storage leak unrelated to this one - the files are there since the cleanup and upgrade and the size and count keeps increasing

8 days after my post is up to 40GB and 576 .sst files (so +16GB and +200 .sst files)

du -hs .local/share/polkadot/chains/ksmcc3/db/full/parachains/db/
40G     .local/share/polkadot/chains/ksmcc3/db/full/parachains/db/

ls -ltr .local/share/polkadot/chains/ksmcc3/db/full/parachains/db/*.sst | wc -l
576

anyone with an active validator can confirm this - I have 2 and both are having the same issue, 2nd one has 40GB and 545 files in the parachains/db folder

**Pull request merged**